File: internal/processor/url_parser.go
// internal/processor/url_parser.go
package processor

import (
	"net/url"
	"regexp"
	"strings"
	"unicode"
)

type URLParser struct {
	shortenerPatterns []*regexp.Regexp
	obfuscationPatterns []*regexp.Regexp
}

type ParsedURL struct {
	Original     string
	Normalized   string
	Scheme       string
	Domain       string
	Subdomain    string
	TLD          string
	Path         string
	QueryParams  map[string]string
	IsShortened  bool
	IsObfuscated bool
	HasIP        bool
}

func NewURLParser() *URLParser {
	return &URLParser{
		shortenerPatterns: []*regexp.Regexp{
			regexp.MustCompile(`bit\.ly`),
			regexp.MustCompile(`tinyurl\.com`),
			regexp.MustCompile(`goo\.gl`),
			regexp.MustCompile(`ow\.ly`),
			regexp.MustCompile(`t\.co`),
			regexp.MustCompile(`buff\.ly`),
		},
		obfuscationPatterns: []*regexp.Regexp{
			regexp.MustCompile(`%[0-9A-Fa-f]{2}`), // URL encoding
			regexp.MustCompile(`@`),               // Userinfo obfuscation
			regexp.MustCompile(`\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}`), // IP address
		},
	}
}

func (p *URLParser) ParseAndAnalyze(rawURL string) (*ParsedURL, error) {
	parsed := &ParsedURL{
		Original:    rawURL,
		QueryParams: make(map[string]string),
	}

	// Normalize URL
	normalized, err := p.normalizeURL(rawURL)
	if err != nil {
		return nil, err
	}
	parsed.Normalized = normalized

	// Parse URL components
	u, err := url.Parse(normalized)
	if err != nil {
		return nil, err
	}

	parsed.Scheme = u.Scheme
	parsed.Domain = u.Hostname()
	parsed.Path = u.Path

	// Extract subdomain and TLD
	parsed.Subdomain, parsed.TLD = p.extractDomainParts(u.Hostname())

	// Parse query parameters
	p.parseQueryParams(u, parsed)

	// Analyze URL characteristics
	parsed.IsShortened = p.isShortenedURL(u)
	parsed.IsObfuscated = p.isObfuscatedURL(u)
	parsed.HasIP = p.containsIPAddress(u.Hostname())

	return parsed, nil
}

func (p *URLParser) normalizeURL(rawURL string) (string, error) {
	// Ensure scheme
	if !strings.Contains(rawURL, "://") {
		rawURL = "https://" + rawURL
	}

	// Parse to validate
	u, err := url.Parse(rawURL)
	if err != nil {
		return "", err
	}

	// Normalize host to lowercase
	u.Host = strings.ToLower(u.Host)

	// Remove default ports
	u.Host = strings.TrimSuffix(u.Host, ":80")
	u.Host = strings.TrimSuffix(u.Host, ":443")

	return u.String(), nil
}

func (p *URLParser) extractDomainParts(host string) (string, string) {
	parts := strings.Split(host, ".")
	if len(parts) < 2 {
		return "", host
	}

	tld := parts[len(parts)-1]
	subdomain := ""

	if len(parts) > 2 {
		subdomain = strings.Join(parts[:len(parts)-2], ".")
	}

	return subdomain, tld
}

func (p *URLParser) parseQueryParams(u *url.URL, parsed *ParsedURL) {
	query := u.Query()
	for key, values := range query {
		if len(values) > 0 {
			parsed.QueryParams[key] = values[0]
		}
	}
}

func (p *URLParser) isShortenedURL(u *url.URL) bool {
	host := u.Hostname()
	for _, pattern := range p.shortenerPatterns {
		if pattern.MatchString(host) {
			return true
		}
	}
	return false
}

func (p *URLParser) isObfuscatedURL(u *url.URL) bool {
	fullURL := u.String()
	for _, pattern := range p.obfuscationPatterns {
		if pattern.MatchString(fullURL) {
			return true
		}
	}
	return false
}

func (p *URLParser) containsIPAddress(host string) bool {
	ipPattern := regexp.MustCompile(`^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$`)
	return ipPattern.MatchString(host)
}

func (p *URLParser) CalculateEntropy(s string) float64 {
	freq := make(map[rune]float64)
	for _, r := range s {
		freq[r]++
	}

	var entropy float64
	length := float64(len(s))
	for _, count := range freq {
		probability := count / length
		entropy -= probability * (probability)
	}

	return entropy
}

func (p *URLParser) DetectHomographAttack(domain string) bool {
	// Check for mixed scripts (Latin + Cyrillic, etc.)
	var hasLatin, hasCyrillic, hasGreek bool

	for _, r := range domain {
		switch {
		case unicode.Is(unicode.Latin, r):
			hasLatin = true
		case unicode.Is(unicode.Cyrillic, r):
			hasCyrillic = true
		case unicode.Is(unicode.Greek, r):
			hasGreek = true
		}
	}

	// If multiple scripts detected, potential homograph attack
	scriptCount := 0
	if hasLatin { scriptCount++ }
	if hasCyrillic { scriptCount++ }
	if hasGreek { scriptCount++ }

	return scriptCount > 1
}

File: netzilla/core/url_parser.py
import re
from typing import Dict, Optional, Tuple
from dataclasses import dataclass, field
from urllib.parse import urlparse, parse_qsl
import unicodedata

@dataclass
class ParsedURL:
    original: str
    normalized: str
    scheme: str
    domain: str
    subdomain: str
    tld: str
    path: str
    query_params: dict[str, str] = field(default_factory=dict)
    is_shortened: bool = False
    is_obfuscated: bool = False
    has_ip: bool = False

class URLParser:
    def __init__(self):
        self.shortener_patterns = [
            re.compile(r'^bit\.ly$'),
            re.compile(r'^tinyurl\.com$'),
            re.compile(r'^goo\.gl$'),
            re.compile(r'^ow\.ly$'),
            re.compile(r'^t\.co$'),
            re.compile(r'^buff\.ly$'),
        ]
        self.obfuscation_patterns = [
            re.compile(r'%[0-9A-Fa-f]{2}'),
            re.compile(r'@'),
            re.compile(r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}'),
        ]
        self.ip_pattern = re.compile(r'^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$')

    def parse_and_analyze(self, raw_url: str) -> ParsedURL | None:
        try:
            normalized = self._normalize_url(raw_url)
            u = urlparse(normalized)
            
            subdomain, tld = self._extract_domain_parts(u.hostname or "")
            
            parsed = ParsedURL(
                original=raw_url,
                normalized=normalized,
                scheme=u.scheme or "",
                domain=u.hostname or "",
                subdomain=subdomain,
                tld=tld,
                path=u.path or "",
                query_params=dict(parse_qsl(u.query))
            )
            
            parsed.is_shortened = self._is_shortened_url(u.hostname or "")
            parsed.is_obfuscated = self._is_obfuscated_url(normalized)
            parsed.has_ip = self._contains_ip_address(u.hostname or "")
            
            return parsed
        except Exception:
            return None

    def _normalize_url(self, raw_url: str) -> str:
        if "://" not in raw_url:
            raw_url = "https://" + raw_url
        
        u = urlparse(raw_url)
        # Normalize host to lowercase
        host = u.hostname.lower() if u.hostname else ""
        
        # Remove default ports
        host = host.replace(":80", "").replace(":443", "")
        
        # Reconstruct URL (simplistic)
        return f"{u.scheme}://{host}{u.path}"

    def _extract_domain_parts(self, host: str) -> tuple[str, str]:
        parts = host.split(".")
        if len(parts) < 2:
            return "", host
        
        tld = parts[-1]
        subdomain = ".".join(parts[:-2]) if len(parts) > 2 else ""
        
        return subdomain, tld

    def _is_shortened_url(self, host: str) -> bool:
        for pattern in self.shortener_patterns:
            if pattern.match(host):
                return True
        return False

    def _is_obfuscated_url(self, full_url: str) -> bool:
        for pattern in self.obfuscation_patterns:
            if pattern.search(full_url):
                return True
        return False

    def _contains_ip_address(self, host: str) -> bool:
        return bool(self.ip_pattern.match(host))

    def calculate_entropy(self, s: str) -> float:
        freq = {}
        for char in s:
            freq[char] = freq.get(char, 0) + 1
        
        entropy = 0.0
        length = float(len(s))
        for count in freq.values():
            prob = count / length
            entropy -= prob * prob
        
        return entropy

    def detect_homograph_attack(self, domain: str) -> bool:
        scripts = set()
        for char in domain:
            try:
                name = unicodedata.name(char)
                if "CYRILLIC" in name:
                    scripts.add("CYRILLIC")
                elif "GREEK" in name:
                    scripts.add("GREEK")
                elif "LATIN" in name:
                    scripts.add("LATIN")
            except ValueError:
                continue
        
        return len(scripts) > 1
